[GTK][WPE][OpenXR] Add a mechanism to exit an immersive experience

spenap opened the pull request:
#### 893ce1bec46e7283563a27396c9cab154770dfbd
<pre>
[GTK][WPE][OpenXR] Add a mechanism to exit an immersive experience
<a href="https://bugs.webkit.org/show_bug.cgi?id=294752">https://bugs.webkit.org/show_bug.cgi?id=294752</a>

Reviewed by Adrian Perez de Castro.

This exposes a mechanism to exit an immersive experience using
PlatformXRSystem::invalidate.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: implement
  webkit_web_view_end_immersive_session by retrieving the XR system, if
any, and invoking invalidate.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in: declare a new
  webkit_web_view_end_immersive_session method.
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::invalidate): use a new InvalidationReason to
decide the next state when invoking invalidateImmersiveSessionState.
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h: add a new
  InvalidationReason enum parameter to invalidate.
* Tools/MiniBrowser/wpe/main.cpp: validate the logic by hooking the
  Escape key with ending an immersive session.

Canonical link: <a href="https://commits.webkit.org/298279@main">https://commits.webkit.org/298279@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8a9c83c5b565d2c806c528919606edf3d1d108a3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114784 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34529 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120946 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116673 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35158 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43087 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87241 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42114 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; layout-tests (exception)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117732 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28010 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103070 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67631 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27195 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21192 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64614 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97394 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124147 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41788 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31217 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96047 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42166 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99260 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95832 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24419 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41014 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18855 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/37843 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41669 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47183 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41231 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42975 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented:
![image](https://github.com/user-attachments/assets/74143e09-e71d-4393-86eb-4625c90b259c)

This approach will currently not work while @svillar is updating the OpenXR implementation, but I actually wonder if it would work after he finishes. 

`xrSystem->invalidate()` is already being called by `WebPageProxy::resetState(ResetStateReason resetStateReason)`, so I guess it's not fundamentally wrong to do it as a WebView-initiated action. 

I've tried to put together a diagram with _what I understand_ would be the sequence. Any comments, @aperezdc ?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/603446c494f88903ef0fd44cce25e83c56fcfda2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46988

spenap commented:
I initially attempted to write a new API test, as suggested by @carlosgcampos. However, what I wrote used an actual PlatformXR instance, so it needed Monado to run.

I ended up spending longer on the test and its environment than on the logic -- and after ending the experience the test would crash. In order to better debug I hooked Minibrowser's Escape key to make it call `webkit_web_view_end_immersive_session`, and this helped me understand that the problem was in invoking `invalidateImmersiveSessionState();` from `PlatformXRSystem::invalidate`, as it transitioned to `ImmersiveSessionState::Idle` and in this case we want it to go into `ImmersiveSessionState::SessionEndingFromSystem`

I have now made that change -- with a boolean argument, which I understand is not the cleanest approach. I'm open to suggestions, @svillar / @aperezdc / @carlosgcampos 

In the meantime I will see if I can use a mock OpenXR implementation in the tests

aperezdc reviewed (approved):
This LGTM from the API design point of view, and the rest looks reasonable, too; but you may want @svillar or someone else with WebXR chops to double check.

I remember that we talked at some point about adding a new GObject property to `WebKitWebView` that would indicate whether an immersive session is active, to allow applications to observe it using the `GObject:notify` signal. Would that go in a follow-up patch?

svillar commented on Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
```diff
@@ -64,7 +64,7 @@ std::optional<SharedPreferencesForWebProcess> PlatformXRSystem::sharedPreference
 }
 
-void PlatformXRSystem::invalidate()
+void PlatformXRSystem::invalidate(bool fromSystem)
```
Basically the same comment

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5843,3 +5844,26 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ *
+ * Since: 2.50
+ */
+void
+webkit_web_view_end_immersive_session(WebKitWebView* webView)
```
```suggestion
void webkit_web_view_end_immersive_session(WebKitWebView* webView)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5843,3 +5844,26 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
 }
+
+/**
+ * webkit_web_view_end_immersive_session:
```
I find a bit weird to have api to end an immersive session when there's no way if there's any immersive session or not, nor a way to start one. I guess the session is started by web content right? In that case I agree with @aperezdc that we should add a property to check if there's a session.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5843,3 +5844,26 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+{
+    g_return_if_fail(WEBKIT_IS_WEB_VIEW(webView));
+
+#if ENABLE(WEBXR)
+    Ref page = getPage(webView);
+    if (auto xrSystem = page->xrSystem()) {
+        // This asks the xr coordinator to end the session for the page it's invoked on,
+        // going through the sessionDidEnd message
+        xrSystem->invalidate(PlatformXRSystem::InvalidationReason::Client);
+    }
+#endif
```
I think this function should return a boolean so that the user knows if the session was finished or not.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -237,6 +237,11 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
         }
     }
 
+    if (keyval == WPE_KEY_Escape) {
+        webkit_web_view_end_immersive_session(webView);
+        return TRUE;
+    }
```
Because here we are claiming that we handle ESC even when there's no session active. If we add the property, we could leave the function as void and require to check the property first.

carlosgcampos reviewed (commented):
I'm sorry to be late here. It's ok to address any issue in a follow up commit anyway.

spenap committed:
[GTK][WPE][OpenXR] Add a mechanism to exit an immersive experience
https://bugs.webkit.org/show_bug.cgi?id=294752

Reviewed by Adrian Perez de Castro.

This exposes a mechanism to exit an immersive experience using
PlatformXRSystem::invalidate.

* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: implement
  webkit_web_view_end_immersive_session by retrieving the XR system, if
any, and invoking invalidate.
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in: declare a new
  webkit_web_view_end_immersive_session method.
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::invalidate): use a new InvalidationReason to
decide the next state when invoking invalidateImmersiveSessionState.
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h: add a new
  InvalidationReason enum parameter to invalidate.
* Tools/MiniBrowser/wpe/main.cpp: validate the logic by hooking the
  Escape key with ending an immersive session.

Canonical link: https://commits.webkit.org/298279@main

webkit-commit-queue commented:
Committed 298279@main (893ce1bec46e): <https://commits.webkit.org/298279@main>

Reviewed commits have been landed. Closing PR #46988 and removing active labels.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -5843,3 +5844,26 @@ webkit_web_view_get_default_content_security_policy(WebKitWebView* webView)
+ * webkit_web_view_end_immersive_session:
+ * @web_view: a #WebKitWebView
+ *
+ * Requests the immersive session associated to this #WebKitWebView to end.
```
We need to at least explain what an immersive session is. (And also how to detect that one has started, when adding the property that Carlos requested.)

mcatanzaro commented:
Please remember that patches that add new public API require two GTK/WPE reviewers. (Sorry, I didn't notice that this was adding a new API.)

My main concern here is that in the past we have added APIs that depend on experimental features, and then never enabled the feature. E.g. `webkit_settings_get_enable_media_stream` was added in 2014 but it didn't do anything until 2022 because `ENABLE(MEDIA_STREAM)` was not enabled until 2022. It really only makes sense to add public API if you have imminent plans to enable `ENABLE(OPENXR)`; if this is not actually planned, then I'd say it's not ready for public API yet. 

In particular, last time we worked on virtual reality in WebKit, we eventually abandoned the project and left the incomplete code in tree for years, so ideally that should not happen again....

aperezdc commented:
> Please remember that patches that add new public API require two GTK/WPE reviewers. (Sorry, I didn't notice that this was adding a new API.)

While he hasn't explicitly r+'d the patch, the API was discussed with @svillar and AFAIU he is onboard. Having a `WebKitWebView` property to let applications observer when an immersive session starts/ends is planned and will come in a follow-up patch, as [discussed in this very PR](https://github.com/WebKit/WebKit/pull/46988#issuecomment-3154980822) 😉 

> My main concern here is that in the past we have added APIs that depend on experimental features, and then never enabled the feature. E.g. `webkit_settings_get_enable_media_stream` was added in 2014 but it didn't do anything until 2022 because `ENABLE(MEDIA_STREAM)` was not enabled until 2022. It really only makes sense to add public API if you have imminent plans to enable `ENABLE(OPENXR)`; if this is not actually planned, then I'd say it's not ready for public API yet.

If this is a concern (is it, for real?) then we can have new experimental API inside `#ifdef` and we can require applications to define `-DWEBKIT_DEFINE_EXPERIMENTAL_API=1` to have headers declare it&mdash;and we can feel free to not consider certain functionality part of the stable API. I am not convinced that we would really want to go this way; but it's an option.

> In particular, last time we worked on virtual reality in WebKit, we eventually abandoned the project and left the incomplete code in tree for years, so ideally that should not happen again....

...which is the reason why now we have been aligning the OpenXR implementation to follow an architecture similar to Apple's, making it work with the GTK port as well, moving code to the UIProcess so XR can work with proper sandboxing, and so on. The idea is to maintain it better going forward, so it doesn't bitrot again.

mcatanzaro commented:
> If this is a concern (is it, for real?)

Well I don't know what your plans are. Currently `ENABLE_WEBXR` is gated behind `ENABLE_EXPERIMENTAL_FEATURES` and historically our experimental build flags have tended to remain experimental for a _very_ long time. If you're planning to enable the feature soon, then this is not a problem. Otherwise, well, it seems much nicer to _not_ add an API that won't do anything.

I'd rather not get into conditional compilation of public APIs, though. That's even worse. I would just keep the API downstream or hidden until you have plans to actually enable the feature.

spenap commented:
Hi @mcatanzaro , @carlosgcampos and @aperezdc , thanks for your discussion. I expect to submit the first follow-up PR shortly, adding `hasImmersiveSession` and addressing your comments. I will tag you all there.

spenap commented:
Follow-up in https://github.com/WebKit/WebKit/pull/49055 for the `hasImmersiveSession` property. Another follow-up is needed for the API test, but I need to complete the mock XR coordinator for that.
